### ISSUE TITLE
Using DOCUMENT() and a sub-query in AQL query causes DOCUMENT() to return NULL

### ISSUE THREAD

**Issue by lostpebble** [arangodb/arangodb#13306]
## My Environment

* __ArangoDB Version__:        3.7.5
* __Storage Engine__:             RocksDB
* __Deployment__:      Oasis (OneShard)

## Background

This issue is based off a post-mortem of a failure to update my deployment on Oasis yesterday from 3.7.3-1 to 3.7.5. After the update I noticed that certain pages on my site were having massive data issues. Upon further investigation, this data issue was caused by AQL queries not returning data like they used to in the previous version.

Today, I created a fresh clone deployment and updated to 3.7.5 again to run tests to see exactly what went wrong.

## Component, Query & Data

On 3.7.5, these two queries return different results:

__AQL query__:
```
LET geoRedirect = DOCUMENT("geolevel_za_url_redirects", "za*cape-town")

LET lockdownCategories = (
  FOR category IN covid_lockdown_categories
  RETURN category
)

return { geoRedirect, lockdownCategories }
```

Even though `geoRedirect` *does* exist for that documenet key and collection, it somehow turns to `null` in the response.

This query works fine:

```
LET geoRedirect = FIRST(FOR redirect IN geolevel_za_url_redirects FILTER redirect._key == "za*cape-town" RETURN redirect)

LET lockdownCategories = (
  FOR category IN covid_lockdown_categories
  RETURN category
)

return { geoRedirect, lockdownCategories }
```

And here are their profiles, respectively:

NOT WORKING:

```
Query String (341 chars, cacheable: false):
 LET geoRedirect = DOCUMENT("geolevel_za_url_redirects", "za*cape-town")
 LET lockdownCategories = (
   FOR category IN covid_lockdown_categories
   RETURN category
 )
 return { geoRedirect, lockdownCategories }

Execution plan:
 Id   NodeType                  Site  Calls   Items   Runtime [s]   Comment
  1   SingletonNode             DBS       1       1       0.00001   * ROOT
  2   CalculationNode           DBS       1       1       0.00005     - LET geoRedirect = DOCUMENT("geolevel_za_url_redirects", "za*cape-town")   /* simple expression */
 11   SubqueryStartNode         DBS       1       2       0.00001     - LET lockdownCategories = ( /* subquery begin */
  4   EnumerateCollectionNode   DBS       1       6       0.00004       - FOR category IN covid_lockdown_categories   /* full collection scan, 1 shard(s) */
 12   SubqueryEndNode           DBS       1       1       0.00006         - RETURN  category ) /* subquery end */
  7   CalculationNode           DBS       1       1       0.00001     - LET #5 = { "geoRedirect" : geoRedirect, "lockdownCategories" : lockdownCategories }   /* simple expression */
  9   RemoteNode                COOR      2       1       0.00057     - REMOTE
 10   GatherNode                COOR      2       1       0.00004     - GATHER   /* unsorted */
  8   ReturnNode                COOR      2       1       0.00001     - RETURN #5

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   cluster-one-shard
  2   splice-subqueries

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           5            0          0         0.00392

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00009
 optimizing ast             0.00001
 loading collections        0.00001
 instantiating plan         0.00003
 optimizing plan            0.00222
 executing                  0.00082
 finalizing                 0.00073
```

WORKING:

```
Query String (341 chars, cacheable: false):
 LET geoRedirect = FIRST(FOR redirect IN geolevel_za_url_redirects FILTER redirect._key == 
 "za*cape-town" RETURN redirect)
 LET lockdownCategories = (
   FOR category IN covid_lockdown_categories
   RETURN category
 )
 return { geoRedirect, lockdownCategories }

Execution plan:
 Id   NodeType                  Site  Calls   Items   Runtime [s]   Comment
  1   SingletonNode             DBS       1       1       0.00001   * ROOT
 21   SubqueryStartNode         DBS       1       2       0.00001     - LET lockdownCategories = ( /* subquery begin */
 10   EnumerateCollectionNode   DBS       1       6       0.00004       - FOR category IN covid_lockdown_categories   /* full collection scan, 1 shard(s) */
 22   SubqueryEndNode           DBS       1       1       0.00001         - RETURN  category ) /* subquery end */
 19   SubqueryStartNode         DBS       1       2       0.00001     - LET #1 = ( /* subquery begin */
 16   IndexNode                 DBS       1       2       0.00005       - FOR redirect IN geolevel_za_url_redirects   /* primary index scan, 1 shard(s) */    
 15   LimitNode                 DBS       1       2       0.00001         - LIMIT 0, 1
 20   SubqueryEndNode           DBS       1       1       0.00005         - RETURN  redirect ) /* subquery end */
 13   CalculationNode           DBS       1       1       0.00002     - LET #10 = { "geoRedirect" : FIRST(#1), "lockdownCategories" : lockdownCategories }   /* simple expression */
 17   RemoteNode                COOR      2       1       0.00066     - REMOTE
 18   GatherNode                COOR      2       1       0.00007     - GATHER   /* unsorted */
 14   ReturnNode                COOR      2       1       0.00001     - RETURN #10

Indexes used:
 By   Name      Type      Collection                  Unique   Sparse   Selectivity   Fields       Ranges
 16   primary   primary   geolevel_za_url_redirects   true     false       100.00 %   [ `_key` ]   (redirect.`_key` == "za*cape-town")

Optimization rules applied:
 Id   RuleName
  1   remove-unnecessary-calculations
  2   optimize-subqueries
  3   use-indexes
  4   remove-filter-covered-by-index
  5   remove-unnecessary-calculations-2
  6   move-calculations-down
  7   cluster-one-shard
  8   splice-subqueries

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           5            1          0         0.00580

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00014
 optimizing ast             0.00001
 loading collections        0.00001
 instantiating plan         0.00004
 optimizing plan            0.00389
 executing                  0.00096
 finalizing                 0.00075
```

I have a feeling this may have to do with OneShard and the new changes in `3.7.5` that allow the optimized use of `DOCUMENT()` in one-shard queries. I did try and run a quick test on the community version on my PC, and it seemed like this issue didn't exist (non-OneShard deployment).

---

On an unrelated note, after updating my deployment and having these issues I was unable to roll back to a previous backup in Oasis. No matter how many times I tried- my deployment was now stuck on this new version. This in turn caused us around 2 hours of down time on our app, and was only solved thanks to some intervention from the ArangoDB engineers (thanks @kvahed and @fceller ). I think this also needs to be urgently looked into.

Of course, in the future I shall now always roll a new deployment for testing before updating the main. I just assumed a patch update had no breaking changes, such as these. But bugs are bound to happen from time to time, understandably. The ability to roll back would have minimized the stress though.

**Comment by fceller** [comment]
Hi @lostpebble we will look into it. This looks like a problem in the optimizer.

**Comment by Simran-B** [comment]
Possibly related to #12994 and #13328.
#13324 also sounds related, although it is said to only occur after an upgrade from 3.6 (unclear if a fresh but cloned Oasis deployment can also be affected).

**Comment by lostpebble** [comment]
This is now fixed in 3.7.6-1, upgraded my deployment and have not found any further issues with it.
